src/emulator/functionsEmulatorRuntime.ts

```diff
@@ -446,61 +450,72 @@ async function initializeFirebaseFunctionsStubs(frb: FunctionsRuntimeBundle): Pr
-
-  httpsProvider[methodName] = (
-    handler: (req: Request, resp: Response) => void,
-    opts: DeploymentOptions
-  ) => {
-    const cf = requestWithOptions(handler, opts);
+  httpsProvider[onRequestInnerMethodName] = (handler: HttpsHandler, opts: DeploymentOptions) => {
+    const cf = onRequestMethodOriginal(handler, opts);
     cf.__emulator_func = handler;
     return cf;
   };
 
-  /*
-    If you take a look at the link above, you'll see that onRequest relies on _onRequestWithOptions
-    so in theory, we should only need to mock _onRequestWithOptions, however that is not the case
-    because onRequest is defined in the same scope as _onRequestWithOptions, so replacing
-    the definition of _onRequestWithOptions does not replace the link to the original function
-    which onRequest uses, so we need to manually force it to use our error-handle-able version.
-     */
-  httpsProvider.onRequest = (handler: (req: Request, resp: Response) => void) => {
-    return httpsProvider[methodName](handler, {});
+  // If you take a look at the link above, you'll see that onRequest relies on _onRequestWithOptions
+  // so in theory, we should only need to mock _onRequestWithOptions, however that is not the case
+  // because onRequest is defined in the same scope as _onRequestWithOptions, so replacing
+  // the definition of _onRequestWithOptions does not replace the link to the original function
+  // which onRequest uses, so we need to manually force it to use our version.
+  httpsProvider.onRequest = (handler: HttpsHandler) => {
+    return httpsProvider[onRequestInnerMethodName](handler, {});
   };
 
-  const onCallOriginal = httpsProvider.onCall;
-  httpsProvider.onCall = (handler: (data: any, context: any) => any | Promise<any>) => {
-    // Look for a special header provided by the functions emulator that lets us mock out
-    // callable functions auth.
-    const newHandler = (data: any, context: any) => {
-      if (context.rawRequest) {
-        const authContext = context.rawRequest.header(HttpConstants.CALLABLE_AUTH_HEADER);
-        if (authContext) {
-          logDebug("Callable functions auth override", {
-            key: HttpConstants.CALLABLE_AUTH_HEADER,
-            value: authContext,
-          });
-          context.auth = JSON.parse(authContext);
-        } else {
-          logDebug("No callable functions auth found");
-        }
-      }
+  // Mocking https.onCall is very similar to onRequest
+  let onCallInnerMethodName = "_onCallWithOpts";
+  if (compareVersionStrings(firebaseFunctionsResolution.version, "3.1.0") >= 0) {
+    onCallInnerMethodName = "_onCallWithOptions";
+  }
+  const onCallMethodOriginal = httpsProvider[onCallInnerMethodName];
 
-      return handler(data, context);
-    };
+  httpsProvider[onCallInnerMethodName] = (handler: CallableHandler, opts: DeploymentOptions) => {
+    const wrapped = wrapCallableHandler(handler);
+    const cf = onCallMethodOriginal(wrapped, opts);
```

Comment:
@abeisgoat I didn't do the `__emulator_func` thing here as that was not compatible with `onCall`.  Just want to double check that's ok.

Reply:
yeah so `__emulator_func` was a hack to get around our implementation of pure HTTPS functions to never returning the result of the user function (because it was wrapped in `(req, res):void => {}`) so since onCall methods support promises natively (like other background functions) then you shouldn't need it, so as long as you verify that returning a long-lived promise works as intended then you should be good to drop it.

Reply:
OK I will give that a try.

Reply:
Tested with this function, worked as expected:
```js
exports.onCallPromise = functions.runWith({
  timeoutSeconds: 60,
  memory: "1GB"
}).https.onCall((data, context) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({ 
        status: 'waited, cool beans',
        auth: context.auth,
      });
    }, 3000);
  });
});
```